File: embeddings-2024-03-11/bunjs/embed.ts
import { Glob } from "bun";
import ollama from "ollama";
import { XMLParser } from "fast-xml-parser";


function timeToSeconds(time: string): number {
  const [hours, minutes, seconds] = time.split(':').map(parseFloat);
  const totalSeconds = hours * 3600 + minutes * 60 + seconds;
  return Math.round(totalSeconds);
}

function chunkSubtitles(subs: subtitleJson[], chunkSize: number, chunkOverlap: number): subtitleJson[] {
  const chunkedSubs: subtitleJson[] = [];
  const maxSubs = subs.length - 1;
  let i = 0;
  while (i < maxSubs - chunkSize) {
    const index = i;
    const start = subs[index].start;
    const text = subs.slice(i, i + chunkSize).reduce((a, b) => { a.text += ` ${b.text}`; return a; }).text;
    chunkedSubs.push({
      start, text, embed: [], videoID: subs[index].videoID, videoTitle: subs[index].videoTitle
    })
    i = i + chunkSize - chunkOverlap;
  }

  return chunkedSubs;
}

function getTitleAndID(filename: string): { title: string, id: string } {
  let title = "";
  let id = "";
  const regex = /^(.*?)\s+\[(.*?)\]\..*?$/;
  const matches = filename.match(regex);
  if (matches && matches.length >= 3) {
    title = matches[1];
    id = matches[2];
  }
  return { title, id }
}


function parseXML(xmlstring: string, id: string, title: string): subtitleJson[] {
  const subtitles: subtitleJson[] = []
  const parser = new XMLParser({ ignoreAttributes: false });
  let st: [] = parser.parse(xmlstring).tt.body.div.p;
  for (const p of st) {
    const text = p["#text"];

    const start = timeToSeconds(p["@_begin"]);
    subtitles.push({ start, text, embed: [], videoID: id, videoTitle: title })
  }
  return subtitles
}

type subtitleJson = {
  start: number;
  text: string;
  embed: number[];
  videoID: string;
  videoTitle: string;
}

const glob = new Glob("*.ttml");
const subtitleFiles: string[] = []

const allSubtitles: subtitleJson[] = [];
for await (const file of glob.scan("./captions")) {
  let seconds = 0
  const { title, id } = getTitleAndID(file)

  console.log("Title:", title);
  console.log("ID:", id);
  const xmlSubtitlesText = (await Bun.file(`./captions/${file}`).text()).replaceAll("<br />", " ");
  const subtitles = parseXML(xmlSubtitlesText, id, title);

  const chunkedSubtitles = chunkSubtitles(subtitles, 10, 5);

  for await (const chunk of chunkedSubtitles) {
    const { embedding } = await ollama.embeddings({ model: 'nomic-embed-text', prompt: chunk.text });
    chunk.embed = embedding
  }
  allSubtitles.push(...chunkedSubtitles);

  const outputfile = "./embeddedSubtitles.json";
  await Bun.write(outputfile, JSON.stringify(allSubtitles, null, 2));

}

File: embeddings-2024-03-11/python/embed.py
import re, glob, ollama, json, os
from typing import List
import xml.etree.ElementTree as ET

def time_to_seconds(time):
  hours, minutes, seconds = [int(x) for x in time.split(":")]
  total_seconds = hours * 3600 + minutes * 60 + seconds
  return total_seconds


def chunk_subtitles(subs, chunk_size, chunk_overlap):
    max_subs = len(subs) - 1
    i = 0
    chunked_subs = []

    while i < (max_subs - chunk_size):
        index = i
        start = subs[index]['start']
        text = ' '.join([sub['text'] for sub in subs[i:i+chunk_size]])

        chunked_subs.append({
            "start": start,
            "text": text,
            "embed": [],
            "videoID": subs[index]["videoID"],
            "videoTitle": subs[index]["videoTitle"]
        })

        i = i + chunk_size - chunk_overlap

    return chunked_subs

def get_title_and_id(filename):
    title = ""
    id = ""
    file = os.path.splitext(os.path.basename(filename))[0]
    regex = r"(.*?)\s+\[(.*?)\]\..*?"
    matches = re.match(regex, file)

    if matches and len(matches.groups()) >= 2:
        title = matches.group(1)
        id = matches.group(2)

    return {'title': title, 'id': id}

def parse_xml(xmlstring, id, title):
    tree = ET.fromstring(xmlstring)
    elements = tree.findall('.//{http://www.w3.org/ns/ttml}p')
    subtitles = []
    for p in elements:
        text = p.text.strip()
        start = p.attrib['begin']
        subtitles.append({"start": start, "text": text, "embed": [], "videoID": id, "videoTitle": title})
    
    return subtitles

if __name__ == "__main__":
    files = glob.glob("../captions/*.ttml")
    all_subtitles = []

    for file in files:
        title, id = get_title_and_id(file).values()
        with open(file, "r") as f:
          xml_subtitles_text = f.read().replace("<br />", "")
          subtitles = parse_xml(xml_subtitles_text, id, title)
          chunked_subtitles = chunk_subtitles(subtitles, 10, 5)

          for chunk in chunked_subtitles:
              embedding = ollama.embeddings(model='nomic-embed-text', prompt = chunk["text"] )['embedding']
              chunk['embed'] = embedding

          all_subtitles += [chunk]

          
          with open('embeddedSubtitles.json','w') as f:
              f.write(json.dumps(all_subtitles, indent=2))              